src/emailagent.cpp

```diff
@@ -228,6 +238,45 @@ bool EmailAgent::ipcConnected()
     return !m_waitForIpc;
 }
 
+bool EmailAgent::isOnline()
+{
+    return m_nmanager->isOnline();
+}
+
+void EmailAgent::searchMessages(const QMailMessageKey &filter,
+                                const QString &bodyText, QMailSearchAction::SearchSpecification spec,
+                                quint64 limit, const QMailMessageSortKey &sort)
+{
+    // Only one search action should be running at time,
+    // cancel any running or queued
+    m_enqueing = true;
+    cancelSearch();
+    qCDebug(lcDebug) << "Enqueuing new search " << bodyText;
+    m_enqueing = false;
+    enqueue(new SearchMessages(m_searchAction.data(), filter, bodyText, spec, limit, sort));
+}
+
+void EmailAgent::cancelSearch()
+{
+    //cancel running action if is search
+    if (m_currentAction && (m_currentAction->type() == EmailAction::Search)) {
+        if (m_currentAction->serviceAction()->isRunning()) {
+            m_cancellingSingleAction = true;
+            m_currentAction->serviceAction()->cancelOperation();
+        } else {
+            m_currentAction = QSharedPointer<EmailAction>();
+            processNextAction();
+        }
+    }
+    // Starts from 1 since top of the queue will be removed if above conditions are met.
+    for (int i = 1; i < m_actionQueue.size(); ++i) {
+        if (m_actionQueue.at(i).data()->type() == EmailAction::Search) {
+            m_actionQueue.removeAt(i);
```

Comment:
This will skip the item at (i+1) since i will be incremented in the loop mutator.